### PR TITLE
fix(database): Improve missing data message

### PR DESCRIPTION
**Pull request by gggritso**
Several important improvements to the "No queries found" banner in the Database overview.

1. The "No queries found" message _only shows up if there's no data on the screen_. Previously, it did it's own check. This was portable, but it was inconsistent. People were seeing the banner even though there's data below. Now, the banner component is told by the parent whether there's data or not
2. Small improvements to copy and UI. Prevents duplicate projects in the lists, extracts some components, improves grammar, removes unused code.
3. New "You have omitted projects" message for those who have been denylisted, so they can contact support. Checks the metrics ingestion flag, since that's a 1:1 proxy for our denylisting.
4. The "No queries found", "You have outdated SDKs" and "Your projects were omitted" messages are now more independent. The omission message _always_ show up if true, even if there's data! The SDK message only shows up if there's no data. The "No data" messages shows up if there's no data

### CODE REVIEW

File: static/app/views/performance/database/noDataMessage.tsx [hunk 1]
```diff
@@ -1,106 +1,117 @@
...
+            ),
+          }
+        )}{' '}
+      {outdatedProjects.length > 0 &&
+        tct('You may be missing data due to outdated SDKs: [projectList].', {
+          projectList: <ProjectList projects={outdatedProjects} />,
+        })}{' '}
+      {denylistedProjects.length > 0 &&
+        tct(
+          'Some of your projects have been omitted from database metrics extraction. Please contact [support]. Omitted projects: [projectList].',
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/noDataMessage.tsx:70]: l: Ideally we can avoid talking about metrics in general.
```suggestion
          'Some of your projects have been omitted from query metrics calculations. Please contact [support]. Omitted projects: [projectList].',
```

File: static/app/views/performance/database/useDenylistedProjects.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * Returns a list of projects that are not eligible for span metrics
+ * because they were denylisted.
+ *
+ * @param options Additional options
+ * @param options.projectId List of project IDs to check against. If omitted, checks all organization projects
+ * @returns List of projects
+ */
+export function useDenylistedProjects(options?: Options) {
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/useDenylistedProjects.tsx:16]: m: Where is `enabled` being used here?

File: static/app/views/performance/database/noDataMessage.tsx [hunk 3]
```diff
@@ -1,106 +1,117 @@
...
+          {
+            support: (
+              <ExternalLink href="https://sentry.io/support/">support</ExternalLink>
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/noDataMessage.tsx:73]: h: Is there a better link to contact support? This goes to the Zendesk help center, but that doesn't allow folks to email in, so kind of confusing workflow. We should prob also give support heads up that this exists.

File: static/app/views/performance/database/noDataMessage.tsx [hunk 4]
```diff
@@ -1,106 +1,117 @@
...
   }
 
-  if (!projectSpanMetricsCounts) {
+  // Hide message if all conditions are satisfied
+  if (
+    isDataAvailable &&
+    !areOutdatedProjectsFetching &&
+    !areDenylistProjectsFetching &&
+    outdatedProjects.length === 0 &&
+    denylistedProjects.length === 0
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/noDataMessage.tsx:48]: l: I'd prefer if we extract out these two conditions (`Hide message while fetching outdated projects or denylisted projects` and `Hide message if all conditions are satisfied) are extract into variables, and we have a single `return null`. This hopefully means that we can get rid of the comments as well.

File: static/app/views/performance/database/useDenylistedProjects.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  const {projects, fetching} = useProjects();
+
+  const shouldCheckAllProjects =
+    options?.projectId?.length === 0 || options?.projectId?.includes('-1');
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/useDenylistedProjects.tsx:20]: l: Isn't it enough to only do `options?.projectId?.includes` check?

Another nit, I find these kind of conditionals hard to read because we have a mixture of `undefined` + `boolean` and coercion is hard to follow. Can we do early returns if `options` or `options.projectId` is `undefined` and then deal with fully validated data later on?

File: static/app/views/performance/database/useDenylistedProjects.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+  const projectsToCheck = shouldCheckAllProjects
+    ? projects
+    : projects.filter(project => options?.projectId?.includes(project.id.toString()));
+
+  const denylistedProjects = projectsToCheck
+    .filter(project => {
+      return !project?.features.includes('span-metrics-extraction');
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/useDenylistedProjects.tsx:28]: is features really conditionally defined on `project`? This feels like something we should fix with the whole codebase tbh.

**Review by AbhiPrasad — COMMENTED** [review]
Not approving because we need to address support link - refactor feels solid though.

File: static/app/views/performance/database/noDataMessage.tsx (hunk 1 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx:70]: I agree but I'm confused because your suggestion still mentions metrics!

File: static/app/views/performance/database/noDataMessage.tsx (hunk 3 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx:73]: Going to ask in Slack!

File: static/app/views/performance/database/useDenylistedProjects.tsx (hunk 2 above)
Comment (gggritso) [reply static/app/views/performance/database/useDenylistedProjects.tsx:16]: Good catch, removed

File: static/app/views/performance/database/useDenylistedProjects.tsx (hunk 6 above)
Comment (gggritso) [reply static/app/views/performance/database/useDenylistedProjects.tsx:28]: It is not, removed

File: static/app/views/performance/database/useDenylistedProjects.tsx (hunk 5 above)
Comment (gggritso) [reply static/app/views/performance/database/useDenylistedProjects.tsx:20]: I don't think that's enough, the app behaves the same whether project ID is explicitly `-1` or missing completely in the URL. There are similar checks in other parts of Performance.

There's also no early return to be done here, since none of the conditions are invalid, they just have different code paths

If you have thoughts on making this easier to read, I'm all ears! I personally find that if I squint and pretend the `?.` aren't there it reads okay. I can assign the conditions to variables, and their name would clarify, maybe?

File: static/app/views/performance/database/noDataMessage.tsx (hunk 4 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx:48]: I totally get the extract into variables part, but why the single `return null`? Does that help readability in any way? Plus, if combined I have to evaluate _both_ conditions before `returning null`. Peanuts performance-wise, but still!

File: static/app/views/performance/database/noDataMessage.tsx (hunk 3 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx:73]: Okay, here's the scoop:

- generally, everywhere in the UI, we direct people to that link. There are lots of examples. The page is good because it's a portal to all kinds of info, and it makes slightly more sense for Self-Hosted folks who might need to read those docs
- I could also run `openHelpSearchModal`, which is what happens when you click "Contact Support" in the sidebar. This is good because it's a more direct link to the Support team, and it opens the Zendesk modal if that's available. This is probably my favourite option here, and Support like it 'cause it appends relevant page context
- some parts of the app just link to email support directly, which I guess also works

File: static/app/views/performance/database/noDataMessage.tsx (hunk 1 above)
Comment (AbhiPrasad) [reply static/app/views/performance/database/noDataMessage.tsx:70]: Sorry should have made it clearer that my suggestion is what I think we should do ideally (since it still mentions metrics), but at least avoids having to bring up extraction.

If we can find a way to remove mentioning metrics all together even better. 

Maybe something like `query performance analysis`, which matches some of the vernacular in the docs.

```suggestion
          'Some of your projects have been omitted from query performance analysis. Please contact [support]. Omitted projects: [projectList].',
```

File: static/app/views/performance/database/noDataMessage.tsx (hunk 3 above)
Comment (AbhiPrasad) [reply static/app/views/performance/database/noDataMessage.tsx:73]: Ideally we actually show something else for self-hosted to direct users to config directly, but I'm not sure if its worth blocking the PR.

I like the `openHelpSearchModal` option too!

File: static/app/views/performance/database/noDataMessage.tsx (hunk 4 above)
Comment (AbhiPrasad) [reply static/app/views/performance/database/noDataMessage.tsx:48]: The single return is more of a consequence of extract outing the variable, but I wouldn't be opposed to keeping it two different return conditions. I just would like the clarity the naming brings.

File: static/app/views/performance/database/useDenylistedProjects.tsx (hunk 5 above)
Comment (AbhiPrasad) [reply static/app/views/performance/database/useDenylistedProjects.tsx:20]: Thinking about this more, rather than early returns, I think default values improves the readability for me.

Something like so:

```ts
export function useDenylistedProjects(options: Options = {}) {
  const {projects, fetching} = useProjects();

  const { projectId = [] } = options;

  const shouldCheckAllProjects = projectId.length === 0 || projectId.includes('-1');
  // ...
```

my opinion is that propagating down `undefined` means I have the mental overhead when reading to understand what happens if optional chaining causes a shortcircuit and coerces the value to be `undefined`.

File: static/app/views/performance/database/noDataMessage.tsx (hunk 1 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx:70]: Works for me!

File: static/app/views/performance/database/useDenylistedProjects.tsx (hunk 5 above)
Comment (gggritso) [reply static/app/views/performance/database/useDenylistedProjects.tsx:20]: 👍🏻 works for me

File: static/app/views/performance/database/noDataMessage.tsx (hunk 3 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx]: Done 👌🏻

File: static/app/views/performance/database/noDataMessage.tsx (hunk 4 above)
Comment (gggritso) [reply static/app/views/performance/database/noDataMessage.tsx]: Makes sense. I split the conditions more and grouped them up to improve meaning, should be better now!

File: static/app/views/performance/database/noDataMessage.tsx [hunk 7]
```diff
@@ -1,106 +1,120 @@
...
+        tct('You may be missing data due to outdated SDKs: [projectList].', {
+          projectList: <ProjectList projects={outdatedProjects} />,
         })}{' '}
+      {denylistedProjects.length > 0 &&
+        tct(
+          'Some of your projects have been omitted from query performance analysis. Please [supportLink]. Omitted projects: [projectList].',
+          {
+            supportLink: (
+              <Button priority="link" onClick={() => openHelpSearchModal({organization})}>
+                {t('contact support')}
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/noDataMessage.tsx:75]: Not sure between `Contact support` and `Contact Support`, just took my best guess at style.

```suggestion
                {t('Contact Support')}
```

File: static/app/views/performance/database/useDenylistedProjects.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+export function useDenylistedProjects(options: Options = {}) {
+  const {projects, fetching} = useProjects();
+
+  const {projectId = []} = options;
+
+  const shouldCheckAllProjects = projectId.length === 0 || projectId.includes('-1');
+
+  const projectsToCheck = shouldCheckAllProjects
+    ? projects
+    : projects.filter(project => options?.projectId?.includes(project.id.toString()));
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/useDenylistedProjects.tsx:24]:
```suggestion
    : projects.filter(project => projectId.includes(project.id.toString()));
```

File: static/app/views/performance/database/useDenylistedProjects.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import useProjects from 'sentry/utils/useProjects';
+
+interface Options {
+  projectId?: string[];
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/useDenylistedProjects.tsx:4]: l: we might want to rename to `projectIds` to make it clear it is an array.

**Review by AbhiPrasad — APPROVED** [review]

File: static/app/views/performance/database/useDenylistedProjects.tsx (hunk 9 above)
Comment (gggritso) [reply static/app/views/performance/database/useDenylistedProjects.tsx:4]: I was torn on this, since most of the app calls it `projectId` despite being plural 😭 keeping as-is for now, but I agree with this complaint!
